### DISCUSSION TITLE
How do I log sensor data on Spike Prime?

### DISCUSSION THREAD

**Discussion by mihye126** [pybricks/support#796] [Q&A]
I found DataLog module in EV3 but it is not working on spike prime. Also I cannot write or create csv file.
How do I log sensor data on Spike Prime?

**Comment by laurensvalk** [comment]
There is an idea to add built-in support for that class in https://github.com/pybricks/support/issues/401, but it isn't selected for implementation yet.

However, you can quite easily do this yourself as follows. First, use a script that prints out comma-separated data, like this:

```python
from pybricks.pupdevices import Motor, UltrasonicSensor
from pybricks.parameters import Port, Direction
from pybricks.robotics import DriveBase
from pybricks.tools import wait, StopWatch

# Initialize devices
left_motor = Motor(Port.A, Direction.COUNTERCLOCKWISE)
right_motor = Motor(Port.B)
sensor = UltrasonicSensor(Port.F)
drive_base = DriveBase(left_motor, right_motor, wheel_diameter=56, axle_track=112)

# Start driving
drive_base.drive(200, 45)

# Start a timer
watch = StopWatch()

for i in range(20):

    # Measure some data
    time = watch.time()
    drive_angle = drive_base.angle()
    drive_distance = drive_base.distance()
    sensor_distance = sensor.distance()

    # Print comma separated row of data
    print(time, drive_angle, drive_distance, sensor_distance, sep=", ")

    # Wait some time between measurements
    wait(100)
```

The data will be printed in the terminal window at the bottom:

![image](https://user-images.githubusercontent.com/12326241/201466556-39a46eb8-98c9-41f3-bd72-37b9b607ddae.png)

Then you can copy it to your favorite spreadsheet tool, which creates columns automatically:

![image](https://user-images.githubusercontent.com/12326241/201466482-4e1ce77b-65ca-4ed0-8967-c8db68cefb3c.png)

And make a graph:

![image](https://user-images.githubusercontent.com/12326241/201466500-0f209264-4a0c-4014-950a-8e026492e7cc.png)

**Comment by dsplaisted** [comment]
Is there a good way to log high-frequency data?  I'd like to get something like [gyro girl](https://www.youtube.com/watch?v=Yc_FEriEmQ8) working, but I may need to sample at a pretty high rate, and it looks like the Pybricks Code truncates the output after about 1000 lines.  It seems like it may be hard to capture the data I need.

Full-fledged support for something like #401 would be great.  It would also be nice if it would work in VS Code, I will probably eventually switch to using that.

**Reply by laurensvalk** [reply]
Here is some [example code](https://github.com/pybricks/pybricks-projects/blob/master/sets/mindstorms-robot-inventor/other-models/balancer/main.py) for [this balancer](https://www.youtube.com/watch?v=muVCJEmVa_M). Contrary to popular belief, PID is not the right control scheme for this sort of thing :-)

A balancing loop doesn't need to run much faster than 50 Hz, so 1000 lines may still give you up to 20 seconds of data, which may be enough depending on what you'd like to do.

> Is there a good way to log high-frequency data?

For higher frequencies, you could store it in an array and print it after the experiment, but it would still be about the same order of magnitude of logging duration I think, since it has up to 250KB of ram and there will be some overhead.

**Reply by dsplaisted** [reply]
Very cool, and that looks way more stable then the other examples I saw.

Is there an explanation somewhere of the algorithm it uses?

It looks somewhat similar to PID to me, the `angle` seems to be the proportional component, and the `rate` would be the derivative.  The main thing I don't really understand is the `speed` component, it looks like it's the difference between the motor position and the motor position 300 ms ago.  It seems like since the bot drives at a constant rate, this element would pretty much stay the same.

**Reply by laurensvalk** [reply]
We've got a few tutorials on the way, but it will be a little while.

It does have parallels to a PID controller, but with a key difference. PID is mainly used to control 1 output variable with 1 input. For example to control motor position using voltage input. In the balancer (viewed from the side; let's disregard steering), there is only one input (motor voltage / duty cycle) but _two_ outputs: robot angle and position. This is generally a lot harder to control and classic PD control doesn't work. A single PID controller may keep it upright but it will just drive off eventually, usually falling over.

But intuitively, you can think of the example above as "two PD controllers" at the same time, adding up to control the same input voltage signal to the motor. The speed component is the "D" term of the "second" PD controller, in effect slowing the robot down, although it achieves this by _speeding up_ (which is another departure from PID, which generally uses negative feedback).
